src/nav_channel.cpp

```diff
@@ -0,0 +1,221 @@
+#include <ros/ros.h>
+#include <nav_channel/TaskStatus.h>
+#include <nav_channel/TaskGoalPosition.h>
+#include <nav_channel/Task.h>
+#include <string>
+//#include <prop_mapper/PropArray.h> // isn't included yet
+#include <nav_channel/PropArray.h> // fake prop map to replace prop_mapper for now
+#include <geometry_msgs/PoseWithCovarianceStamped.h>
+#include <nav_msgs/Odometry.h>
+
+class NavChannel {
+public:
+    NavChannel(): nh("")
+    {
+        // task_to_exec will confirm we are on nav_channel, then execute task
+        task_to_exec = nh.subscribe("task_to_execute", 10, &NavChannel::navChannelCallback, this);
+        
+        // props will get the prop map and store it
+        prop_map = nh.subscribe("props", 10, &NavChannel::propMapCallback, this);
+
+        // global_position/local grabs position from mavros and stores it
+        global_pos = nh.subscribe("mavros/global_position/local", 10, &NavChannel::globalPositionCallback, this);
+
+        // we publish to task_status to update us on progress of task
+        task_status = nh.advertise<nav_channel::TaskStatus>("task_status", 10);
+
+        // we publish to task_goal_position to send the boat to a waypoint
+        task_goal_position = nh.advertise<nav_channel::TaskGoalPosition>("task_goal_position", 10);
+    }
+
+    void spin() {
+        ros::Rate rate(10);
+        while (ros::ok()) {
+            ros::spinOnce();
+            rate.sleep();
+        }
+    }
+
+    void setDestination(geometry_msgs::PoseStamped midpoint) {
+        // sets goal_pos and then publishes it
+        goal_pos.goal_pose = midpoint;
+        // will set orientation to (0,0,0,0), so probably want to grab current
+        // orientation and set that to new orientation
+
+        ROS_INFO_STREAM("Midpoint set at " << midpoint.pose.position.x << ","<< midpoint.pose.position.y << "," << midpoint.pose.position.z);
+
+        task_goal_position.publish(goal_pos);
+    }
+
+    geometry_msgs::PoseStamped findMidpoint() {
+        ROS_INFO("in findMidpoint.");
+        // validates positions of props, then calculates midpoint, and returns
+        // it as PoseStamped
+        geometry_msgs::PoseStamped midpoint;
+
+        bool red = false;   // use two booleans to determine if props exist
+        bool green = false; // add math to determine if valid postions
+        nav_channel::Prop red_prop;     // again fake props, change later
+        nav_channel::Prop green_prop;
+
+        for (int i = 0; !red || !green; i++) {
+            if (props.props[i].prop_label == "Red Prop") {
```

Comment:
We should probably add some checks in to make sure it's selecting the closest red and green props, It could pick up stuff farther in the distance first.